src/jit/lsraxarch.cpp

```diff
@@ -360,47 +438,85 @@ void LinearScan::BuildNode(GenTree* tree)
         case GT_TEST_EQ:
         case GT_TEST_NE:
         case GT_CMP:
-            BuildCmp(tree);
+            srcCount = BuildCmp(tree);
             break;
 
         case GT_CKFINITE:
-            appendLocationInfoToList(tree->gtOp.gtOp1);
-            info->srcCount = 1;
-            assert(info->dstCount == 1);
-            info->internalIntCount = 1;
-            break;
+        {
+            assert(dstCount == 1);
+            RefPosition* internalDef = buildInternalIntRegisterDefForNode(tree);
+            srcCount                 = BuildOperandUses(tree->gtOp.gtOp1);
+            buildInternalRegisterUses();
+            BuildDef(tree);
+        }
+        break;
 
         case GT_CMPXCHG:
         {
-            info->srcCount = 3;
-            assert(info->dstCount == 1);
-
-            // comparand is preferenced to RAX.
-            // Remaining two operands can be in any reg other than RAX.
-            LocationInfoListNode* locationInfo = getLocationInfo(tree->gtCmpXchg.gtOpLocation);
-            locationInfo->info.setSrcCandidates(this, allRegs(TYP_INT) & ~RBM_RAX);
-            useList.Append(locationInfo);
-            LocationInfoListNode* valueInfo = getLocationInfo(tree->gtCmpXchg.gtOpValue);
-            valueInfo->info.setSrcCandidates(this, allRegs(TYP_INT) & ~RBM_RAX);
-            useList.Append(valueInfo);
-            info->setDstCandidates(this, RBM_RAX);
-            LocationInfoListNode* comparandInfo = getLocationInfo(tree->gtCmpXchg.gtOpComparand);
-            comparandInfo->info.setSrcCandidates(this, RBM_RAX);
-            useList.Append(comparandInfo);
+            srcCount = 3;
+            assert(dstCount == 1);
+
+            // Comparand is preferenced to RAX.
+            // The remaining two operands can be in any reg other than RAX.
+            BuildUse(tree->gtCmpXchg.gtOpLocation, allRegs(TYP_INT) & ~RBM_RAX);
+            BuildUse(tree->gtCmpXchg.gtOpValue, allRegs(TYP_INT) & ~RBM_RAX);
+            BuildUse(tree->gtCmpXchg.gtOpComparand, RBM_RAX);
+            BuildDef(tree, RBM_RAX);
         }
         break;
 
+        case GT_XADD:
+        case GT_XCHG:
         case GT_LOCKADD:
-            info->srcCount = appendBinaryLocationInfoToList(tree->AsOp());
-            assert(info->dstCount == (tree->TypeGet() == TYP_VOID) ? 0 : 1);
-            break;
+        {
+            // TODO-XArch-Cleanup: We should make the indirection explicit on these nodes so that we don't have
+            // to special case them.
+            // These tree nodes will have their op1 marked as isDelayFree=true.
+            // That is, op1's reg remains in use until the subsequent instruction.
+            GenTree* addr = tree->gtOp.gtOp1;
```

Comment:
I would suggest using functions like `gtGetOp1` and `gtGetOp2` consistently

Reply:
Will do; I was programmed to avoid it back before we added `gtGetOp2IfPresent()`, and I guess I'm a creature of habit.